marvelai-org/marvel-ai-backend
# Pull Request - CI/CD Part 1 Dynamic Env - Digital Dreamers

## Summary
The .env file with the utility function configures environment variables during local development. This allows developers to customize their local environment without affecting the production settings. 
## Changes

- Two files added: `env_manager.py` (Location: app/services/env_manager.py) and `test_env_manager.py` (Location: app/services/tests/test_env_manager.py)

- Changes made to the `./local-start.sh` to incorporate the utility function
<img width="739" alt="Screenshot 2024-08-31 at 11 13 59" src="https://github.com/user-attachments/assets/43e7fb9f-2e3a-4ca9-92ae-8aa18265f4f0">


- Changes made to the `.gitignore` to ignore the folder keys/
- Change made in the `requirements.txt`, fastapi changed to fastapi[standard]


## Testing
The new feature was tested using the following unit tests and integration tests. Refer to the documentation for more details. 

### I) Unit Tests

* ‘Test_env_variable_from_env_file’： this test verifies that the ‘get_env_variable’ function correctly retrieves environment variables from a ‘.env’ file.
* ‘test_env_variable _from_google_secrets’: this test checks that the ‘get_env_variable’ function correctly retrieve environment variables from Google Cloud’s Secret Manager when the variable is not found in the ‘.env’ file.

**Steps to run the unit tests:**

Execute the test command: `pytest app/services/tests/test_env_manager.py`
Review the test output: after running the command, you’ll see output indicating whether the tests passed or failed. 
Expected Output: Upon successful execution, you should see the following output in terminal:
<img width="1079" alt="Screenshot 2024-09-07 at 12 03 21" src="https://github.com/user-attachments/assets/fd54267c-f60a-49e1-8855-0c9d1d78c501">


### II) Integration Tests
* The integration test ensures that the ‘get_env_variable’ function works correctly when integrated into the broader application environment. This test verifies that environment variables are correctly loaded from both the ‘.env’ file and Google Cloud’s Secret Manager, and that the application starts with the correct configuration. 

**Steps to run the integration test:**
Run the Startup Script: 

```
./local-start.sh
```
The script begins by setting up local environment variables required by the application. 
It authenticates the Google Cloud project using the service account credentials specified in the ‘local-auth.json’ file.
The ‘get_env_variable’ function is called to fetch environment variables, ‘GOOGLE_API_KEY’, ‘ENV_TYPE’，and ‘PROJECT_ID’. These variables are either retrieved from the local ‘.env’ file or from Google Cloud Secret Manager if not found locally.
Expected Output: Upon successful execution, you should see the following output in terminal:
<img width="1083" alt="Screenshot 2024-08-31 at 11 16 32" src="https://github.com/user-attachments/assets/3e2ddca1-66e2-4da6-b39b-adfab88241fc">



## Results
The new feature passed the above-mentioned tests and performed its required function without breaking the existing codebase. 
## Notes
* The utility function ‘get_env_varible’ is designed to streamline the process of retrieving environment variables from both the .env file (for local development) and Google Cloud Secrets Manager (for production).
* Documentation: [CI/CD Part : Dynamic Environment](https://docs.google.com/document/d/1VZQl_bsYYF2rXquBGTakuMlw3GIfsiJvO8hWj_yyR9E/edit?usp=sharing)
* This function attempts to retrieve environment variables in a secure and reliable manner. It first attempts to retrieve the environment variable from a local .env file. If the variable is not found there, the function then tries to retrieve it from Google Cloud Secret Manager.
## Screenshots
The utility function with relevant error handling performed.
<img width="1045" alt="Screenshot 2024-08-31 at 10 13 06" src="https://github.com/user-attachments/assets/278babde-56cd-4a54-8921-c29fbecaae13">
<img width="973" alt="Screenshot 2024-08-31 at 10 13 22" src="https://github.com/user-attachments/assets/615edc9b-85ef-4f40-b540-8212ea8bf7ec">
